test/end2end_test.go

```diff
@@ -7032,6 +7032,84 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 	}()
 }
 
+// TestClientTransportRestartsAfterStreamIdExhausted tests that the client transport
+// drains and restarts when next stream ID exceeds MaxStreamID. This test also
+// verifies that subsequent RPCs use a new client transport and the old transport
+// is closed.
+func (s) TestClientTransportRestartsAfterStreamIdExhausted(t *testing.T) {
+	// overriding MaxStreamIDForTesting.
+	originalMaxStreamID := transport.MaxStreamIDForTesting
+	transport.MaxStreamIDForTesting = 5
+	defer func() {
+		transport.MaxStreamIDForTesting = originalMaxStreamID
+	}()
+
+	// setting up StubServer.
+	s := grpc.NewServer()
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			for i := 0; i < 10; i++ {
+				m := &grpc.PreparedMsg{}
```

Comment:
Why a `PreparedMsg`? That seems unnecessarily complex.  Just do `stream.SendMsg(testpb.WhateverThisIs{})`.